### ISSUE TITLE
CascadeMode improvements

### ISSUE THREAD

**Issue by JeremySkinner** [FluentValidation/FluentValidation#1437]
`CascadeMode.StopOnFirstFailure` only affects rules in the same `RuleFor` chain, even when set at the validator level. This causes confusion:

```csharp
CascadeMode = CascadeMode.StopOnFirstFailure;
RuleFor(x => x.Property).NotNull().NotEqual("bar");
RuleFor(x => x.AnotherProperty).NotNull().NotEqual("baz");
```

...so if the NotNull fails on the first rule, then the NotEqual will not run. However, the second rule will *still run* because these are distinct/independent rule chains. IE, this is the equivalent of:

```csharp
RuleFor(x => x.Property).Cascade(CascadeMode.StopOnFirstFailure).NotNull().NotEqual("bar");
RuleFor(x => x.AnotherProperty).Cascade(CascadeMode.StopOnFirstFailure).NotNull().NotEqual("baz");
```

This is by design, but it isn't intuitive and often causes confusion. I don't think changing this behaviour is a good idea as it would be an extremely subtle breaking change that wouldn't be caught by the compiler. Even with a major version bump this is still a very subtle change. 

Investigate alternative approaches, which might involve adding new options to the CascadeMode enum to enable this behaviour.

For example, could add a new enum entry `CascadeMode.FailFast`. This would do the same as StopOnFirstFailure, plus also bail out after the first failure from *any* rule.

**Comment by weitzhandler** [comment]
> I don't think changing this behaviour is a good idea as it would be an extremely subtle breaking change that wouldn't be caught by the compiler.

Agreed 100%. The documentation should probably be elaborated, but the functionality should be left intact.

> Investigate alternative approaches, which might involve adding new options to the CascadeMode enum to enable this behaviour.
> 
> For example, could add a new enum entry `CascadeMode.FailFast`. This would do the same as StopOnFirstFailure, plus also bail out after the first failure from _any_ rule.

That's a great idea.

We could choose to mark `StopOnFirstFailure` as `Obsolete` and add another more self-explanatory enum member of an equal value, accompanying the naming style of its fail-fast/short-circuit counterpart.

**Comment by weitzhandler** [comment]
Feel free to assign me, I can take this.
Would appreciate some general guidance on how you want me to implement this.

**Comment by weitzhandler** [comment]
Also how to name the enum fields and whether we want to obsolete the existing one(s).

**Comment by JeremySkinner** [comment]
Thanks for offering to take this on!

Currently `PropertyRule.Validate` and `PropertyRule.ValidateAsync` check if the cascade is set to `StopOnFirstFailure`. If this is set and the rule has a failure, then the rule stops executing. These will also need to be updated to also check for `FailFast`.

Similar logic will need to be introduced in `AbstractValidator.Validate(context)` and `AbstractValidator.ValidateAsync(context, token)`. If the cascade mode at the validator level is set to `FailFast`, then immediately stop executing rules if there's at least one failure from the current rule. Currently the synchronous overload executes the rules using a linq statement, lets change this to a loop (like ValidateAsync has), which will make it more readable.

Tests will also need to be added (the `CascadingFailureTester` is probably the best class for this). 

> Also how to name the enum fields and whether we want to obsolete the existing one(s).

I think let's add just 1 new one for now (`FailFast`, unless you have another name preference). And yes let's obsolete `StopOnFirstFailure` (can introduce a new equivalent of this down the line if we want to).

**Comment by weitzhandler** [comment]
How about 

```c#
[Obsolete("Use FailChain blah blah")]
StopOnFirstFailure,
FailChain = StopOnFirstFailure,
FailRuleSet,
FailAll
```

**Comment by weitzhandler** [comment]
Actually, "fail" might not be the matching word from a "cascade" point of view, which this enum is called, so we could consider `None`, `Chain`, `RuleSet`, `All`, where `All = Continue` and `None` means `FailFast`.

Example: `Cascade.All`, `Cascade.None`. I think this it explains itself better.

**Comment by JeremySkinner** [comment]
I think All/None is quite opaque and hard to understand just from the names. I also don't want to introduce any additional options at this stage.

**Comment by JeremySkinner** [comment]
Maybe just Stop? So let's start with

```
public Enum CascadeMode {
  Continue,
  [Obsolete]
  StopOnFirstFailure,
  Stop,
}
```

Please don't introduce any new options related to ruleset at this point.

**Comment by JeremySkinner** [comment]
This is a chain (a single rule, with multiple validators):
```
RuleFor(x => x.Foo).NotNull().NotEqual("foo");
```

This is a ruleset (a grouping of multiple rules):

```
RuleSet("foo", () => {
   RuleFor(x => x.Foo).NotNull();
   RuleFor(x => x.Bar).NotNull();
});
```

**Comment by weitzhandler** [comment]
OK, so I understood it well about the difference between chain and set.

Anyway, if we add `Stop`, `StopOnFirstFailure` gives a connotation of a more strict fail-fast execution. So if we're obsoleting `StopOnFirstFailure`, we should provide a different name with the same values this is better self-explanatory.

**Comment by weitzhandler** [comment]
> Please don't introduce any new options related to rulesets at this point.

Wouldn't we want to include those in the future? If we introduce `Stop` now, without accommodating for rule-sets upfront it will lead to more confusion then.
Alright enough chattery, please lemme know your final enum structure, including what we replace `StopOnFirstFailure` with (because it's a useful functionality in its own right, just has a confusing name), and I'll start coding.

Thanks for bearing with me!

**Comment by weitzhandler** [comment]
Alternatively, I'm thinking we could change the `RuleSet()` methods to return an `IRuleSetBuilderInitial<T>`, which will use the same cascade mode features as `RuleFor`. If this is what we're aiming at, `Continue` and `Stop` makes perfect sense.

**Comment by JeremySkinner** [comment]
> Alternatively, I'm thinking we could change the `RuleSet()` methods to return an `IRuleSetBuilderInitial<T>`

That is too large a change for me to be thinking about/reviewing at the moment as I only have limited time to work on FluentValidation, so if we can keep this focused on just the single/simple change then this is far more to make it into the library. 

I’ll have a think about what I want the option to be called over the weekend. if you want to make a start before then then I’d suggest just pick something and can rename it after.

**Comment by weitzhandler** [comment]
> That is too large a change for me to be thinking about/reviewing at the moment as I only have limited time to work on FluentValidation, so if we can keep this focused on just the single/simple change then this is far more to make it into the library.

So all we're doing here is a simple rename from `StopOnFirstFailure` to `Stop`? I mean I can do it, but that's probably wasting time for both of us.
Please lemme know if/when I could commit to a bigger change that will include cascading/uncascading between rulesets. I'd be happy to contribute that with tests.

> I’ll have a think about what I want the option to be called over the weekend. if you want to make a start before then then I’d suggest just pick something and can rename it after.

I rather go for the whole nine yards (or at least basic next-rule set policy - stop/continue) instead of spending time on cosmetic changes.

**Comment by JeremySkinner** [comment]
No, as per my original issue description the purpose here is to add a new option to the enum which enables cascading between separate rule chains when this option is used at the validator level. This is a new piece of functionality - not a rename - but is not related to rulesets. Does that make sense?

If that isn't something you're interested in working on then that's fine, but please let me know now.

If you want to discuss cascading between rulesets, we can do that, but that isn't the goal here. We can open a separate issue to discuss and track that, but it's a larger issue that needs a separate discussion.

**Comment by weitzhandler** [comment]
Oh I understand. I actually understood it the other day and managed to reconfuse myself!

> If you want to discuss cascading between rulesets, we can do that, but that isn't the goal here. We can open a separate issue to discuss and track that, but it's a larger issue that needs a separate discussion.

Yes, definitely, that's what I'm aiming for!

> If that isn't something you're interested in working on then that's fine, but please let me know now.

The difference between chained and unchained validations are not much of a difference to me.  
I might find time to work on it over the coming weeks, but it's much lower priority than rulesets.

**Comment by JeremySkinner** [comment]
Implemented in #1478

**Comment by JeremySkinner** [comment]
9.1 is now available on nuget with this change.

**Comment by massada** [comment]
@JeremySkinner 

Just want to point out that the deprecation warning is telling us to change to a wrong value.

[Obsolete("Use CascadeMode.StopFast instead. The behaviour of StopOnFirstFailure was ambiguous when used at validator-level. For more details, see https://docs.fluentvalidation.net/en/latest/conditions.html#stop-vs-stoponfirstfailure")]

Should be CascadeMode.Stop

**Comment by JeremySkinner** [comment]
Thanks, I’ll fix that in a patch release 👍

**Comment by JeremySkinner** [comment]
Fixed in 9.1.1
